### PR TITLE
collective.xkey backports

### PR DESCRIPTION
**Pull request by mamico**
Improvements ported from collective.xkey

The helper view has been changed from the one in `collective.xkey`, because calling a view that removes cached content should be avoided  to users, due to possible DoS.

### CODE REVIEW

File: src/collective/purgebyid/events.py [hunk 1]
```diff
@@ -1,4 +1,5 @@
 # -*- coding: utf-8 -*-
+from collective.purgebyid import logger
 from collective.purgebyid.api import getInvolvedObjs
 from collective.purgebyid.api import markInvolvedObjs
```
Comment (pgrunewald) [inline src/collective/purgebyid/events.py]: I would use `mark_involved_objects` and not the old style.

File: src/collective/purgebyid/events.py [hunk 2]
```diff
@@ -62,11 +58,7 @@ def mutate(self):
 
 @adapter(IPubAfterTraversal)
 def handle_request_after_traversal(event):
-    """handle "IPubAfterTraversal".
-
-    TODO: all the objects traversed are involved or only the last (the
-          firstest within request.PARENTS)?
-    """
+    """handle "IPubAfterTraversal"."""
```
Comment (pgrunewald) [inline src/collective/purgebyid/events.py:61]: There is a (maybe a more conceptual) difference to c.xkey: I also mark the published content as being involved as the user sees this content as well. Is there a reason why it is commented out?

**Review by pgrunewald — CHANGES_REQUESTED** [review]
Overall the PR looks very good. There are though some minor things I have found.

File: src/collective/purgebyid/events.py (hunk 2 above)
Comment (mamico) [reply src/collective/purgebyid/events.py:61]: I have to double-check, but if I remember correctly, the first `parent` of the request is the published context itself. Am I wrong?

https://github.com/collective/collective.purgebyid/commit/b7610b8270aac6ac764fb606c53da02fe9ca2d13

File: src/collective/purgebyid/events.py (hunk 1 above)
Comment (mamico) [reply src/collective/purgebyid/events.py]: changed.

File: src/collective/purgebyid/events.py (hunk 2 above)
Comment (pgrunewald) [reply src/collective/purgebyid/events.py:61]: I inspected the results for a simple request (Plone/front-page). The first parent is not the same as the published object. However the published content is usually the view responsible for rendering (which contains no IUUID) and the first parent is typically the actual content we are interested in. So this just works fine.

As a side note when calling a simple file from a resource directory, that FilesystemFile is the published content and the FilesystemResourceDirectory is the first parent. I'm not sure if we need to handle those cases anyway, because files do not need to be purged? In the end, this is a topic out of scope for this PR.

tl;dr LGTM

**Review by pgrunewald — APPROVED** [review]
